### PR TITLE
Text API: WIP

### PR DESCRIPTION
**Pull request by Defman**
Feedback: Am I going the right direction?

### CODE REVIEW

File: core/src/text.rs [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    #[serde(rename = "show_entity")] 
+    // TODO: Entity struct
+    ShowEntity(String),
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum TextValue {
+    Text {
+        text: String,
```
Comment (caelunshun) [inline core/src/text.rs]: Could we use `Cow<'static, str>` to allow either static strings without allocation or heap-allocated `String`s? Functions which take an `&str` could then be switched to take `impl Into<Cow<'static, str>>`.

**Comment by caelunshun** [comment]
This looks great, thanks for the dedicated work! I think you're on the right path here—I can't see anything which could become a problem in the future.

**Comment by Defman** [comment]
Current only `TextComponent` support the builder patter, ie. TextComponent.red() should `Text` also support this? Current `Text` supports `Text::from("hello world") * Color::Red`.

**Comment by Defman** [comment]
The translate enum should also contain all translation keys for 1.13.2.

**Comment by caelunshun** [comment]
Just did a thorough review—phenomenal work. A few suggestions/questions:
* Could we add some docs which explain the differences between `Text`, `TextComponent`, `TextValue`, and `TextRoot`? I feel this could be somewhat confusing for new developers.
* Also, there seem to be two ways to build `Text` objects: using the operator overloading style API and the `TextComponentBuilder` API. What are the key differences between these two designs? If they both do the same thing, could we decide on just one of them, for simplicity?

**Comment by Defman** [comment]
Operator overloading is supposed to be a wrapper around `TextComponentBuilder`, such that
```rust
impl<T> std::ops::Mul<Color> for T
where
    T: TextComponentBuilder,
```

**Pull request merged**